File: data/src/main/java/com/github/anrimian/musicplayer/data/database/dao/play_list/PlayListsDaoWrapper.java
package com.github.anrimian.musicplayer.data.database.dao.play_list;

import static com.github.anrimian.musicplayer.data.database.utils.DatabaseUtils.getSearchArg;
import static com.github.anrimian.musicplayer.data.database.utils.DatabaseUtils.getSearchArgs;

import androidx.sqlite.db.SimpleSQLiteQuery;

import com.github.anrimian.musicplayer.data.database.LibraryDatabase;
import com.github.anrimian.musicplayer.data.database.dao.compositions.CompositionsDao;
import com.github.anrimian.musicplayer.data.database.entities.playlist.PlayListEntryEntity;
import com.github.anrimian.musicplayer.data.models.exceptions.DuplicatePlaylistEntriesException;
import com.github.anrimian.musicplayer.data.models.exceptions.PlayListAlreadyExistsException;
import com.github.anrimian.musicplayer.data.repositories.scanner.storage.playlists.m3uparser.PlayListEntry;
import com.github.anrimian.musicplayer.data.storage.providers.playlists.AppPlayList;
import com.github.anrimian.musicplayer.data.storage.providers.playlists.StoragePlayList;
import com.github.anrimian.musicplayer.data.storage.providers.playlists.StoragePlayListItem;
import com.github.anrimian.musicplayer.domain.models.composition.Composition;
import com.github.anrimian.musicplayer.domain.models.playlist.PlayList;
import com.github.anrimian.musicplayer.domain.models.playlist.PlayListItem;
import com.github.anrimian.musicplayer.domain.utils.functions.Function;

import java.util.ArrayList;
import java.util.Date;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

import javax.annotation.Nullable;

import io.reactivex.rxjava3.core.Observable;

public class PlayListsDaoWrapper {

    private final PlayListDao playListDao;
    private final CompositionsDao compositionsDao;
    private final LibraryDatabase libraryDatabase;

    public PlayListsDaoWrapper(PlayListDao playListDao,
                               CompositionsDao compositionsDao,
                               LibraryDatabase libraryDatabase) {
        this.playListDao = playListDao;
        this.compositionsDao = compositionsDao;
        this.libraryDatabase = libraryDatabase;
    }

    public void insertStoragePlaylist(StoragePlayList playList, List<StoragePlayListItem> entries) {
        libraryDatabase.runInTransaction(() -> {
            long storageId = playList.getStorageId();
            if (playListDao.isPlayListExistsByStorageId(storageId)) {
                return;
            }

            long id = playListDao.insertPlayList(
                    storageId,
                    getUniquePlayListName(playList.getName()),
                    playList.getDateAdded(),
                    playList.getDateModified()
            );
            insertPlayListItems(entries, id);
        });
    }

    public long insertPlayList(String name,
                               Date dateAdded,
                               Date dateModified,
                               List<Long> compositionsIds) {
        return libraryDatabase.runInTransaction(() -> {
            long playlistId = playListDao.insertPlayList(null, name, dateAdded, dateModified);
            insertPlayListEntries(playlistId, compositionsIds);
            return playlistId;
        });
    }

    public void setPlayListEntries(long playlistId, List<Long> compositionsIds) {
        libraryDatabase.runInTransaction(() -> {
            if (!playListDao.isPlaylistExists(playlistId)) {
                return;
            }
            playListDao.clearPlayListEntries(playlistId);
            insertPlayListEntries(playlistId, compositionsIds);
        });
    }

    public long insertPlayList(String name,
                               Date dateAdded,
                               Date dateModified,
                               Function<Long> storagePlayListFetcher) {
        if (playListDao.isPlayListWithNameExists(name)) {
            throw new PlayListAlreadyExistsException();
        }
        Long storageId = storagePlayListFetcher.call();
        long id = playListDao.insertPlayList(storageId, name, dateAdded, dateModified);
        if (id == -1) {
            throw new IllegalStateException("db not modified");
        }
        return id;
    }

    public List<AppPlayList> getAllAsStoragePlayLists() {
        return playListDao.getAllAsStoragePlayLists();
    }

    public List<AppPlayList> getAllPlayLists() {
        return playListDao.getAllPlayLists();
    }

    public AppPlayList getPlayList(long playlistId) {
        return playListDao.getPlayList(playlistId);
    }

    public void deletePlayList(long id) {
        playListDao.deletePlayList(id);
    }

    public void updatePlayListName(long id, String name) {
        if (playListDao.isPlayListWithNameExists(name)) {
            throw new PlayListAlreadyExistsException();
        }
        playListDao.updatePlayListName(id, name);
    }

    public Observable<List<PlayList>> getPlayListsObservable(String searchQuery) {
        return playListDao.getPlayListsObservable(getSearchArg(searchQuery));
    }

    public Observable<PlayList> getPlayListsObservable(long id) {
        return playListDao.getPlayListObservable(id)
                .takeWhile(items -> !items.isEmpty())
                .map(items -> items.get(0));
    }

    public Observable<List<PlayListItem>> getPlayListItemsObservable(long playListId,
                                                                     boolean useFileName,
                                                                     @Nullable String searchText) {
        String query = PlayListDao.getPlaylistItemsQuery(useFileName);
        Object[] args = new Object[4];
        args[0] = playListId;
        String[] searchArgs = getSearchArgs(searchText, 3);
        System.arraycopy(searchArgs, 0, args, 1, 3);
        SimpleSQLiteQuery sqlQuery = new SimpleSQLiteQuery(query, args);
        return playListDao.getPlayListItemsObservable(sqlQuery);
    }

    public Set<Long> getPlayListsForCompositions(List<Long> compositionIds) {
        Set<Long> playlistsIds = new HashSet<>();
        for (long compositionId: compositionIds) {
            playlistsIds.addAll(playListDao.getPlaylistsForComposition(compositionId));
        }
        return playlistsIds;
    }

    public List<Long> getCompositionIdsInPlaylist(long playlistId) {
        return playListDao.getCompositionIdsInPlaylist(playlistId);
    }

    public List<Composition> getCompositionsInPlaylist(long playlistId, boolean useFileName) {
        String query = PlayListDao.getCompositionsQuery(useFileName);
        SimpleSQLiteQuery sqlQuery = new SimpleSQLiteQuery(query, new Object[] { playlistId });
        return playListDao.getCompositionsInPlaylist(sqlQuery);
    }

    public List<PlayListEntry> getPlayListItemsAsFileEntries(long playListId) {
        return playListDao.getPlayListItemsAsFileEntries(playListId);
    }

    public int deletePlayListEntry(long id, long playListId) {
        return libraryDatabase.runInTransaction(() -> {
            int position = playListDao.selectPositionById(id);
            playListDao.deletePlayListEntry(id);
            playListDao.decreasePositionsAfter(position, playListId);
            return position;
        });
    }

    public void insertPlayListItems(List<StoragePlayListItem> items, long playListId) {
        insertPlayListItems(items, playListId, playListDao.selectMaxOrder(playListId));
    }

    //? change checkForDuplicates, ignoreDuplicates to enum
    // DuplicateCheckType: INSERT_ALL, INSERT_EXCLUDE_DUPLICATES, ABORT_ON_DUPLICATE
    public List<Composition> addCompositions(List<Composition> compositions,
                                             long playListId,
                                             int position,
                                             boolean checkForDuplicates,
                                             boolean ignoreDuplicates) {
        List<Composition> compositionsToInsert = compositions;
        if (checkForDuplicates || ignoreDuplicates) {
            List<Composition> duplicates = getCompositionsInPlaylist(playListId, compositions);
            if (!duplicates.isEmpty()) {
                if (!ignoreDuplicates) {
                    boolean hasNonDuplicates = duplicates.size() < compositions.size();
                    throw new DuplicatePlaylistEntriesException(duplicates, hasNonDuplicates);
                }
                compositionsToInsert = new ArrayList<>(compositions.size());
                Set<Composition> duplicatesSet = new HashSet<>(duplicates);
                for (Composition composition : compositions) {
                    if (!duplicatesSet.contains(composition)) {
                        compositionsToInsert.add(composition);
                    }
                }
            }
        }
        addCompositions(compositionsToInsert, playListId, position);
        return compositionsToInsert;
    }

    public void insertPlayListItems(List<StoragePlayListItem> items,
                                    long playListId,
                                    int position) {
        libraryDatabase.runInTransaction(() -> {
            playListDao.increasePositionsByCountAfter(items.size(), position, playListId);

            List<PlayListEntryEntity> entities = new ArrayList<>(items.size());
            int orderPosition = position;
            for (StoragePlayListItem item : items) {
                long compositionId = compositionsDao.selectIdByStorageId(item.getAudioId());
                if (compositionId == 0) {
                    continue;
                }

                PlayListEntryEntity entryEntity = new PlayListEntryEntity(
                        item.getItemId(),
                        compositionId,
                        playListId,
                        orderPosition++
                );
                entities.add(entryEntity);
            }
            playListDao.insertPlayListEntries(entities);
        });
    }

    public Long selectStorageId(long id) {
        return playListDao.selectStorageId(id);
    }

    public Long selectStorageItemId(long id) {
        return playListDao.selectStorageItemId(id);
    }

    public void moveItems(long playListId, int fromPos, int toPos) {
        playListDao.moveItems(playListId, fromPos, toPos);
    }

    public String selectPlayListName(long playListId) {
        return playListDao.selectPlayListName(playListId);
    }

    public long findPlaylist(String name) {
        return playListDao.findPlaylist(name);
    }

    public int getPlaylistSize(long playListId) {
        return playListDao.getPlaylistSize(playListId);
    }

    private void insertPlayListEntries(long playlistId, List<Long> compositionsIds) {
        for (int i = 0; i < compositionsIds.size(); i++) {
            long compositionId = compositionsIds.get(i);
            if (compositionsDao.isCompositionExists(compositionId)) {
                playListDao.insertPlayListEntry(null, compositionId, playlistId, i);
            }
        }
    }

    private void addCompositions(List<Composition> compositions,
                                 long playListId,
                                 int position) {
        libraryDatabase.runInTransaction(() -> {
            playListDao.increasePositionsByCountAfter(compositions.size(), position, playListId);

            List<PlayListEntryEntity> entities = new ArrayList<>(compositions.size());
            int orderPosition = position;
            for (Composition item : compositions) {
                PlayListEntryEntity entryEntity = new PlayListEntryEntity(
                        null,
                        item.getId(),
                        playListId,
                        orderPosition++
                );
                entities.add(entryEntity);
            }
            playListDao.insertPlayListEntries(entities);
            playListDao.updatePlayListModifyTime(playListId, new Date());
        });
    }

    private List<Composition> getCompositionsInPlaylist(long playListId, List<Composition> compositions) {
        List<Long> playlistCompositions = playListDao.getCompositionIdsInPlaylist(playListId);
        Set<Long> playlistCompositionsSet = new HashSet<>(playlistCompositions);
        List<Composition> foundCompositions = new ArrayList<>();
        for (Composition composition: compositions) {
            if (playlistCompositionsSet.contains(composition.getId())) {
                foundCompositions.add(composition);
            }
        }
        return foundCompositions;
    }

    private String getUniquePlayListName(String name) {
        return getUniquePlayListName(name, "");
    }

    private String getUniquePlayListName(String name, String salt) {
        String uniqueName = name;
        int i = 0;
        while (playListDao.isPlayListWithNameExists(uniqueName)) {
            i++;
            uniqueName = name + "("+ i + ")" + salt;
        }
        return uniqueName;
    }

}

File: data/src/main/java/com/github/anrimian/musicplayer/data/database/dao/playlist/PlaylistsDaoWrapper.kt
package com.github.anrimian.musicplayer.data.database.dao.playlist

import androidx.sqlite.db.SimpleSQLiteQuery
import com.github.anrimian.musicplayer.data.database.LibraryDatabase
import com.github.anrimian.musicplayer.data.database.dao.compositions.CompositionsDao
import com.github.anrimian.musicplayer.data.database.entities.playlist.PlaylistEntryEntity
import com.github.anrimian.musicplayer.data.database.utils.DatabaseUtils.getSearchArg
import com.github.anrimian.musicplayer.data.database.utils.DatabaseUtils.getSearchArgs
import com.github.anrimian.musicplayer.data.models.exceptions.DuplicatePlaylistEntriesException
import com.github.anrimian.musicplayer.data.models.exceptions.PlaylistAlreadyExistsException
import com.github.anrimian.musicplayer.data.repositories.playlists.models.PlaylistEntryPosition
import com.github.anrimian.musicplayer.data.storage.providers.playlists.AppPlaylist
import com.github.anrimian.musicplayer.data.storage.providers.playlists.StoragePlaylist
import com.github.anrimian.musicplayer.data.storage.providers.playlists.StoragePlaylistItem
import com.github.anrimian.musicplayer.domain.models.composition.Composition
import com.github.anrimian.musicplayer.domain.models.composition.CompositionModel
import com.github.anrimian.musicplayer.domain.models.order.Order
import com.github.anrimian.musicplayer.domain.models.order.OrderType
import com.github.anrimian.musicplayer.domain.models.playlist.Playlist
import com.github.anrimian.musicplayer.domain.models.playlist.PlaylistEntry
import com.github.anrimian.musicplayer.domain.utils.rx.firstListItemOrComplete
import io.reactivex.rxjava3.core.Observable

class PlaylistsDaoWrapper(
    private val playListDao: PlaylistDao,
    private val compositionsDao: CompositionsDao,
    private val libraryDatabase: LibraryDatabase
) {

    fun insertStoragePlaylist(playList: StoragePlaylist, entries: List<StoragePlaylistItem>) {
        libraryDatabase.runInTransaction {
            val storageId = playList.storageId
            if (playListDao.isPlayListExistsByStorageId(storageId)) {
                return@runInTransaction
            }
            val id = playListDao.insertPlaylist(
                storageId,
                getUniquePlayListName(playList.name),
                playList.addedTime,
                playList.modifiedTime
            )
            insertPlayListItems(entries, id)
        }
    }

    fun insertPlaylist(
        name: String,
        addedTime: Long,
        modifiedTime: Long,
        compositionsIds: List<Long>
    ): Long {
        return libraryDatabase.runInTransaction<Long> {
            val playlistId = playListDao.insertPlaylist(null, name, addedTime, modifiedTime)
            insertPlayListEntries(playlistId, compositionsIds)
            playlistId
        }
    }

    fun setPlayListEntries(playlistId: Long, compositionsIds: List<Long>) {
        libraryDatabase.runInTransaction {
            if (!playListDao.isPlaylistExists(playlistId)) {
                return@runInTransaction
            }
            playListDao.clearPlayListEntries(playlistId)
            insertPlayListEntries(playlistId, compositionsIds)
        }
    }

    fun insertPlaylist(
        name: String,
        addedTime: Long,
        modifiedTime: Long,
        storagePlayListFetcher: () -> Long?
    ): Long {
        if (playListDao.isPlayListWithNameExists(name)) {
            throw PlaylistAlreadyExistsException()
        }
        val storageId = storagePlayListFetcher()
        val id = playListDao.insertPlaylist(storageId, name, addedTime, modifiedTime)
        check(id != -1L) { "db not modified" }
        return id
    }

    fun getAllAsStoragePlayLists(): List<AppPlaylist> {
        return playListDao.getAllAsStoragePlayLists()
    }

    fun getAllPlayLists(): List<AppPlaylist> {
        return playListDao.getAllPlayLists()
    }

    fun getPlayList(playlistId: Long): AppPlaylist? {
        return playListDao.getPlayList(playlistId)
    }

    fun deletePlayList(id: Long) {
        playListDao.deletePlaylist(id)
    }

    fun updatePlayListName(id: Long, name: String) {
        if (playListDao.isPlayListWithNameExists(name)) {
            throw PlaylistAlreadyExistsException()
        }
        playListDao.updatePlaylistName(id, name)
    }

    fun getPlayListsObservable(searchQuery: String?): Observable<List<Playlist>> {
        return playListDao.getPlayListsObservable(getSearchArg(searchQuery))
    }

    fun getPlayListsObservable(id: Long): Observable<Playlist> {
        return playListDao.getPlayListObservable(id)
            .firstListItemOrComplete()
    }

    fun getPlayListItemsObservable(
        playListId: Long,
        useFileName: Boolean,
        searchText: String?
    ): Observable<List<PlaylistEntry>> {
        val query = PlaylistDao.getPlaylistEntriesQuery(useFileName)
        val args = arrayOf<Any?>(playListId, *getSearchArgs(searchText, 3))
        val sqlQuery = SimpleSQLiteQuery(query, args)
        return playListDao.getPlayListItemsObservable(sqlQuery)
    }

    fun getPlayListsForCompositions(compositionIds: List<Long>): Set<Long> {
        val playlistsIds = HashSet<Long>()
        for (compositionId in compositionIds) {
            playlistsIds.addAll(playListDao.getPlaylistsForComposition(compositionId))
        }
        return playlistsIds
    }

    fun getCompositionIdsInPlaylist(playlistId: Long): List<Long> {
        return playListDao.getCompositionIdsInPlaylist(playlistId)
    }

    fun getCompositionsInPlaylist(
        playlistId: Long,
        useFileName: Boolean
    ): List<Composition> {
        val query = PlaylistDao.getCompositionsQuery(useFileName)
        val sqlQuery = SimpleSQLiteQuery(query, arrayOf<Any>(playlistId))
        return playListDao.getCompositionsInPlaylist(sqlQuery)
    }

    fun getPlayListItemsAsFileEntries(playListId: Long): List<com.github.anrimian.musicplayer.data.repositories.scanner.storage.playlists.m3uparser.PlayListEntry> {
        return playListDao.getPlayListItemsAsFileEntries(playListId)
    }

    fun getNextOrderPosition(playListId: Long) = playListDao.selectNextOrderPosition(playListId)

    fun deletePlayListEntry(id: Long, playListId: Long): Int {
        return libraryDatabase.runInTransaction<Int> {
            val position = playListDao.selectPositionById(id)
            playListDao.deletePlayListEntry(id)
            playListDao.decreasePositionsAfter(position, playListId)
            position
        }
    }

    //? change checkForDuplicates, ignoreDuplicates to enum
    // DuplicateCheckType: INSERT_ALL, INSERT_EXCLUDE_DUPLICATES, ABORT_ON_DUPLICATE
    fun addCompositions(
        compositions: List<CompositionModel>,
        playListId: Long,
        position: Int,
        checkForDuplicates: Boolean,
        ignoreDuplicates: Boolean
    ): List<CompositionModel> {
        var compositionsToInsert = compositions
        if (checkForDuplicates || ignoreDuplicates) {
            val duplicates = getCompositionsInPlaylist(playListId, compositions)
            if (duplicates.isNotEmpty()) {
                if (!ignoreDuplicates) {
                    val hasNonDuplicates = duplicates.size < compositions.size
                    throw DuplicatePlaylistEntriesException(duplicates, hasNonDuplicates)
                }
                val duplicatesSet = duplicates.toSet()
                compositionsToInsert = compositions.filterNot { duplicatesSet.contains(it) }
            }
        }
        addCompositions(compositionsToInsert, playListId, position)
        return compositionsToInsert
    }

    fun insertPlayListItems(
        items: List<StoragePlaylistItem>,
        playListId: Long,
        position: Int = playListDao.selectNextOrderPosition(playListId)
    ) {
        libraryDatabase.runInTransaction {
            playListDao.increasePositionsByCountAfter(items.size, position, playListId)
            val entities = ArrayList<PlaylistEntryEntity>(items.size)
            var orderPosition = position
            for (item in items) {
                val compositionId = compositionsDao.selectIdByStorageId(item.audioId)
                if (compositionId == 0L) {
                    continue
                }
                val entryEntity = PlaylistEntryEntity(
                    0,
                    item.itemId,
                    compositionId,
                    playListId,
                    orderPosition++
                )
                entities.add(entryEntity)
            }
            playListDao.insertPlayListEntries(entities)
        }
    }

    fun selectStorageId(id: Long): Long? {
        return playListDao.selectStorageId(id)
    }

    fun selectStorageItemId(id: Long): Long? {
        return playListDao.selectStorageItemId(id)
    }

    fun moveItems(playListId: Long, fromPos: Int, toPos: Int) {
        playListDao.moveItems(playListId, fromPos, toPos)
    }

    fun getEntryPositions(playlistId: Long): List<PlaylistEntryPosition> {
        return playListDao.getEntryPositions(playlistId)
    }

    fun updateEntryPosition(itemId: Long, position: Int) {
        playListDao.updateEntryPosition(itemId, position)
    }

    fun sortEntries(playlistId: Long, order: Order, useFileName: Boolean) {
        val sortOrder = if (order.isReversed) "DESC" else "ASC"
        val orderClause = when (order.orderType) {
            OrderType.NAME -> {
                if (useFileName) "compositions.fileName COLLATE NOCASE $sortOrder"
                else "CASE WHEN compositions.title IS NULL OR compositions.title = '' THEN compositions.fileName ELSE compositions.title END COLLATE NOCASE $sortOrder"
            }
            OrderType.FILE_NAME -> "compositions.fileName COLLATE NOCASE $sortOrder"
            OrderType.DURATION -> "compositions.duration $sortOrder"
            OrderType.ARTIST -> "(SELECT name FROM artists WHERE id = compositions.artistId) COLLATE NOCASE $sortOrder"
            else -> throw IllegalStateException("unknown order type: $order")
        }
        val query = SimpleSQLiteQuery("""
            SELECT play_lists_entries.itemId 
            FROM play_lists_entries 
            INNER JOIN compositions ON play_lists_entries.audioId = compositions.id
            WHERE play_lists_entries.playListId = ?
            ORDER BY $orderClause
        """, arrayOf(playlistId))
        val sortedIds = playListDao.getEntryIdsSorted(query)

        libraryDatabase.runInTransaction {
            sortedIds.forEachIndexed { index, itemId ->
                playListDao.updateEntryPosition(itemId, index)
            }
            playListDao.updatePlayListModifyTime(playlistId, System.currentTimeMillis())
        }
    }

    fun selectPlayListName(playListId: Long): String? {
        return playListDao.selectPlayListName(playListId)
    }

    fun findPlaylist(name: String): Long {
        return playListDao.findPlaylist(name)
    }

    fun getPlaylistSize(playListId: Long): Int {
        return playListDao.getPlaylistSize(playListId)
    }

    private fun insertPlayListEntries(playlistId: Long, compositionsIds: List<Long>) {
        compositionsIds.forEachIndexed { i, compositionId ->
            if (compositionsDao.isCompositionExists(compositionId)) {
                playListDao.insertPlayListEntry(null, compositionId, playlistId, i)
            }
        }
    }

    private fun addCompositions(
        compositions: List<CompositionModel>,
        playListId: Long,
        position: Int
    ) {
        libraryDatabase.runInTransaction {
            playListDao.increasePositionsByCountAfter(compositions.size, position, playListId)
            val entities = ArrayList<PlaylistEntryEntity>(compositions.size)
            var orderPosition = position
            for (item in compositions) {
                val entryEntity = PlaylistEntryEntity(
                    0,
                    null,
                    item.id,
                    playListId,
                    orderPosition++
                )
                entities.add(entryEntity)
            }
            playListDao.insertPlayListEntries(entities)
            playListDao.updatePlayListModifyTime(playListId, System.currentTimeMillis())
        }
    }

    private fun getCompositionsInPlaylist(
        playListId: Long,
        compositions: List<CompositionModel>
    ): List<CompositionModel> {
        val playlistCompositionsSet = playListDao.getCompositionIdsInPlaylist(playListId).toSet()
        return compositions.filter { playlistCompositionsSet.contains(it.id) }
    }

    private fun getUniquePlayListName(name: String, salt: String = ""): String {
        var uniqueName = name
        var i = 0
        while (playListDao.isPlayListWithNameExists(uniqueName)) {
            i++
            uniqueName = "$name($i)$salt"
        }
        return uniqueName
    }
}
